### ISSUE TITLE
Regression between 1.1.1 and 3.0.0 beta 1 when using RSA_NO_PADDING

### ISSUE THREAD

**Issue by dskern-github** [openssl/openssl#16056]
**TL; DR:** In OpenSSL 1.1.1, `EVP_PKEY_CTX_set_rsa_padding (pCtx, RSA_NO_PADDING)` can be used to perform raw RSA encryption and decryption with both public and private keys. In OpenSSL 3.0 beta1, the exact same calling code returns error 0x1C8000A8 "invalid padding mode" when trying to encrypt with a private key or decrypt with a public key. 

**The long form:**

I have been gradually preparing an application that is currently leveraging OpenSSL 1.1.1 for an upgrade to OpenSSL 3.0. In order to support legacy data, in addition to the typical modern cryptographic operations this application also supports a "pre-standard" RSA padding mode that predates the publication of PKCS#1.

In OpenSSL 1.1.1 and earlier, this application used `RSA_public_encrypt`, `RSA_private_decrypt`, `RSA_private_encrypt`, and `RSA_public_decrypt` with `RSA_NO_PADDING` to efficiently perform raw RSA operations. Since those low-level RSA functions have been deprecated in OpenSSL 3.0, I've migrated the implementation over to use the EVP_PKEY equivalents instead. Following recommendations found online and in the latest version of the documentation, `EVP_PKEY_encrypt`, `EVP_PKEY_decrypt`, and `EVP_PKEY_sign` neatly replace the first three low-level RSA calls.  The documentation inaccurately states that `EVP_PKEY_verify` replaces `RSA_public_decrypt`, but `EVP_PKEY_verify_recover` is actually needed instead.

This replacement strategy worked well against OpenSSL 1.1.1, but after swapping in the header files and binaries from OpenSSL 3.0 beta 1 and rebuilding, `EVP_PKEY_CTX_set_rsa_padding` fails with error 0x1C8000A8 (invalid padding mode) when trying to perform a "raw" (RSA_NO_PADDING) RSA encrypt with a private key or a "raw" (RSA_NO_PADDING) RSA decrypt with a public key.

The same set of errors occurs after switching to the OpenSSL 3.0 calling paradigm by explicitly loading the default, legacy, and fips providers and using `EVP_PKEY_CTX_new_from_pkey` instead of `EVP_PKEY_CTX_new`.

**Comment by dskern-github** [comment]
Hand-applied changes from #16068 to my downloaded beta 1 source tarball after a "nmake clean"; reconfigured and currently rebuilding on W64.

**Comment by dskern-github** [comment]
All of my automated RSA tests passed on w64 when using #16068 as a drop-in replacement for OpenSSL 1.1.1, so using the default provider.  The only bad padding-type error was the expected return from a negative test that attempts to "sign" with a private key using OAEP padding, which is patently impossible.

I'll test with the fips provider next.

**Comment by dskern-github** [comment]
All of my automated RSA and ECDSA tests passed on w64 when using #16068 with the new OpenSSL 3.0 calling patterns and "fips=yes" after adding conditionals to use "provider=default" for RSA keys less than 2048 bits and when signing (but not hashing or verifying) with SHA-1.

**Comment by dskern-github** [comment]
Cloned current master to a macOS system, modified with #16068, built, and tested both as a drop-in replacement for OpenSSL 1.1.1 and with the OpenSSL 3.0 calling patterns as above. All of my automated RSA and ECDSA tests passed on macOS.
